### PR TITLE
Add internal IRT implementation of execv()/execve()

### PR DESCRIPTION
**Pull request by alyptik**
Add `execv()` and `execve()` system call implementations to the internal runtime.

Implementation Details:
- Copy the argument and environment vectors from the "parent" NaClApp.
- Duplicate its `cage_id` along with any other relevant internal state.
- Leverage the existing fork() implementation to emulate replacement of the entire process image.

### CODE REVIEW

File: src/trusted/service_runtime/nacl_globals.h [hunk 1]
```diff
@@ -24,6 +24,7 @@ EXTERN_C_BEGIN
 #define PROT_RW (NACL_ABI_PROT_READ | NACL_ABI_PROT_WRITE)
 #define PROT_RX (NACL_ABI_PROT_READ | NACL_ABI_PROT_EXEC)
 #define F_ANON_PRIV (NACL_ABI_MAP_PRIVATE | NACL_ABI_MAP_ANONYMOUS)
+#define SHM_SIZE (1u << 13)
```
Comment (JustinCappos) [inline src/trusted/service_runtime/nacl_globals.h]: why?  Also, will this always be a constant?

File: src/trusted/service_runtime/sel_ldr_standard.c [hunk 2]
```diff
@@ -727,11 +725,11 @@ int NaClCreateMainThread(struct NaClApp     *nap,
    * data.  We are assuming that the caller is non-adversarial and the
    * code does not look like string data....
    */
-  for (i = 0; i < argc && argv; ++i) {
+  for (i = 0; i < argc && argv && argv[i]; ++i) {
```
Comment (JustinCappos) [inline src/trusted/service_runtime/sel_ldr_standard.c]: Why ++i instead of i++?

File: src/trusted/service_runtime/nacl_globals.h (hunk 1 above)
Comment (alyptik) [reply src/trusted/service_runtime/nacl_globals.h]: Whoops, that was leftover code from something else I had planned but decided wasn't worth pursuing! Good catch, removing.

File: src/trusted/service_runtime/sel_ldr_standard.c (hunk 2 above)
Comment (alyptik) [reply src/trusted/service_runtime/sel_ldr_standard.c]: The rest of the `for()` loops in this file use `++i`, I figured it'd be good to stay consistent.

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 3]
```diff
@@ -3965,38 +3965,32 @@ int32_t NaClSysPipe2(struct NaClAppThread *natp, uint32_t *pipedes, int flags) {
...
+  nap_child->running = 0;
   ret = nap_child->cage_id;
-  if (!NaClCreateMainForkThread(nap, natp, nap_child, child_argc, child_argv, NULL)) {
+
+  /* start fork thread */
+  if (!NaClCreateMainForkThread(nap, natp, nap_child, child_argc, child_argv, nap_child->clean_environ)) {
     NaClLog(1, "%s\n", "[NaClSysFork] forking program failed!");
-    ret = -NACL_ABI_ENOMEM;
-    goto out;
+    ret = -NACL_ABI_ENOEXEC;
```
Comment (moyix) [inline src/trusted/service_runtime/nacl_syscall_common.c]: Why the change to ENOEXEC here? I don't think that's a standard error for `fork` (it is for exec, e.g. in case the thing you're trying to exec isn't a properly formatted executable).

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 4]
```diff
@@ -456,62 +456,88 @@ int32_t NaClSysNameService(struct NaClAppThread *natp,
...
-  fd_cage_table[nap->cage_id][newfd] = fd_cage_table[nap->cage_id][oldfd];
-  ret = newfd;
-  nap->fd++;
+  old_desc = fd_cage_table[nap->cage_id][oldfd];
+  if (!(old_nd = NaClGetDesc(nap, old_desc))) {
+    ret= -NACL_ABI_EBADF;
+    goto out;
+  }
+  new_desc = NaClSetAvail(nap, old_nd);
+  fd_cage_table[nap->cage_id][newfd] = new_desc;
```
Comment (moyix) [inline src/trusted/service_runtime/nacl_syscall_common.c:529]: Does `newfd` need to first be closed if it's open (per the `dup2` man page)? Or is that done by `NaClSetAvail` somehow?

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 5]
```diff
@@ -456,62 +456,88 @@ int32_t NaClSysNameService(struct NaClAppThread *natp,
...
 
-done:
+out:
   return ret;
 }
 
 int32_t NaClSysDup3(struct NaClAppThread  *natp,
                     int                   oldfd,
                     int                   newfd,
                     int                   flags) {
```
Comment (moyix) [inline src/trusted/service_runtime/nacl_syscall_common.c:540]: As a general point, can `dup` and `dup2` be implemented as calls to `dup3` with the appropriate flags?

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 6]
```diff
@@ -2792,7 +2790,7 @@ int32_t NaClSysImcRecvmsg(struct NaClAppThread         *natp,
   recv_hdr.iov_length = kern_nanimh.iov_length;
 
   recv_hdr.ndescv = new_desc;
-  recv_hdr.ndesc_length = NACL_ARRAY_SIZE(new_desc);
+  recv_hdr.ndesc_length = sizeof new_desc / sizeof *new_desc;
```
Comment (moyix) [inline src/trusted/service_runtime/nacl_syscall_common.c]: Would prefer to always use parentheses with sizeof – makes things a bit easier to read and removes ambiguity.

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 7]
```diff
@@ -2881,7 +2879,7 @@ int32_t NaClSysImcRecvmsg(struct NaClAppThread         *natp,
 cleanup:
   /* copy out updated desc count, flags */
   if (retval < 0) {
-    for (i = 0; i < NACL_ARRAY_SIZE(new_desc); ++i) {
+    for (i = 0; i < sizeof new_desc / sizeof *new_desc; ++i) {
```
Comment (moyix) [inline src/trusted/service_runtime/nacl_syscall_common.c]: Parentheses again

File: src/trusted/service_runtime/nacl_syscall_common.c (hunk 3 above)
Comment (alyptik) [reply src/trusted/service_runtime/nacl_syscall_common.c]: After checking the manpage and POSIX standard, it seems you are right and `ENOEXEC` isn't even a valid `errno`. Changing it back to `ENOMEM`.

File: src/trusted/service_runtime/nacl_syscall_common.c (hunk 5 above)
Comment (alyptik) [reply src/trusted/service_runtime/nacl_syscall_common.c:540]: Hm, well `dup()` not really, since it returns an arbitrary file descriptor, but I think it may be possible with `dup2()`/`dup3`. I'll see if I can rework `dup3()` to just call `dup2()`.

File: src/trusted/service_runtime/nacl_syscall_common.c (hunk 4 above)
Comment (alyptik) [reply src/trusted/service_runtime/nacl_syscall_common.c:529]: Good catch, I completely forget to close `newfd`. Fixing.

File: src/trusted/service_runtime/nacl_syscall_common.c (hunk 6 above)
Comment (alyptik) [reply src/trusted/service_runtime/nacl_syscall_common.c]: Sure, I can do that.

File: src/trusted/service_runtime/nacl_syscall_common.c (hunk 7 above)
Comment (alyptik) [reply src/trusted/service_runtime/nacl_syscall_common.c]: Will fix.

File: src/trusted/service_runtime/nacl_syscall_common.c (hunk 5 above)
Comment (alyptik) [reply src/trusted/service_runtime/nacl_syscall_common.c:540]: Duplicated code removed.

**Comment by alyptik** [comment]
Thank you for the quick review, merged.
